jsapp/js/account/accountFieldsEditor.component.tsx

```diff
@@ -0,0 +1,281 @@
+import React from 'react';
+import Checkbox from '../components/common/checkbox';
+import TextBox from '../components/common/textBox';
+import {addRequiredToLabel} from 'js/textUtils';
+import envStore from '../envStore';
+import styles from './accountFieldsEditor.module.scss';
+import KoboSelect from 'js/components/common/koboSelect';
+import type {UserFieldName} from './account.constants';
+
+const genderSelectOptions = [
+  {value: 'male', label: t('Male')},
+  {value: 'female', label: t('Female')},
+  {value: 'other', label: t('Other')},
+];
+
+type UserFieldValue = string | boolean;
+
+export interface AccountFieldsValues {
+  name: string;
+  organization: string;
+  organization_website: string;
+  sector: string;
+  gender: string;
+  bio: string;
+  city: string;
+  country: string;
+  require_auth: boolean;
+  twitter: string;
+  linkedin: string;
+  instagram: string;
+}
+
+export type AccountFieldsErrors = {[name in UserFieldName]?: string};
+
+interface AccountFieldsEditorProps {
+  /**
+   * A list of fields to display in editor. Regardless of this list, all
+   * the fields values will be returned with `onChange` callback (to avoid
+   * losing data). If this is not provided, we display all fields :)
+   */
+  displayedFields?: UserFieldName[];
+  /** Errors to be displayed for fields */
+  errors?: AccountFieldsErrors;
+  /**
+   * We need values for all fields, even if only few are displayed (via
+   * `displayedFields` prop)
+   */
+  values: AccountFieldsValues;
+  onChange: (fields: AccountFieldsValues) => void;
+}
+
+/**
+ * A component that displays fields from user account and allows editing their
+ * values. It DOES NOT handle the API calls to update the values on the endpoint.
+ */
+export default function AccountFieldsEditor(props: AccountFieldsEditorProps) {
+  if (!envStore.isReady) {
+    return null;
+  }
+
+  const metadata = envStore.data.getUserMetadataFieldsAsSimpleDict();
+
+  /** Get label and (required) for a given user metadata fieldname */
+  function getLabel(fieldName: UserFieldName): string {
+    const label =
+      metadata[fieldName]?.label ||
+      (console.error(`No label for fieldname "${fieldName}"`), fieldName);
+    const required = metadata[fieldName]?.required || false;
+    return addRequiredToLabel(label, required);
+  }
+
+  function onAnyFieldChange(
+    fieldName: UserFieldName,
+    newValue: UserFieldValue
+  ) {
+    const newValues = {...props.values, [fieldName]: newValue};
+    props.onChange(newValues);
+  }
+
+  /**
+   * Field will be displayed if it's enabled on Back end and it's not omitted
+   * in `displayedFields`.
+   */
+  function isFieldToBeDisplayed(name: UserFieldName) {
+    return (
+      name in metadata &&
+      (!props.displayedFields || props.displayedFields.includes(name))
```

Comment:
This boolean chain could be condensed a bit using `props.displayedFields?.includes(name)`. But that's a small thing.

Reply:
If there are no `props.displayedFields`, I need this to return `true`. If I use `props.displayedFields?.includes(name)` AFAIU this would work differently. I've added helpful comments :)